packages/database/src/changeHandlers/SurveyResponseOutdater.js

```diff
@@ -3,74 +3,264 @@
 
-    if (!this.scheduledUpdatePromise) {
-      this.scheduledUpdatePromise = new Promise(resolve => {
-        this.scheduledUpdatePromiseResolve = resolve;
-      });
+  const comparableResponses = surveyResponses.map(response => ({
+    ...response,
+    // If other ordering criteria are the same, not outdated responses take precedence over outdated
+    // Assign numeric value for comparison
+    notOutdated: response.outdated ? 0 : 1,
+  }));
+
+  return orderBy(
+    comparableResponses,
+    ['end_time', 'notOutdated', 'id'],
+    ['desc', 'desc', 'desc'],
+  )[0].id;
+};
+
+const getDimensionCombo = (surveyResponse, survey) => {
+  const { period_granularity: periodGranularity } = survey;
+  const { data_time: dataTime } = surveyResponse;
+  const { startDate, endDate } = getPeriodRange(periodGranularity, dataTime);
+
+  return {
+    surveyId: survey.id,
+    entityId: surveyResponse.entity_id,
+    startDate,
+    endDate,
+  };
+};
+
+const groupResponsesForSurveyByDimensionCombo = (surveyResponses, survey) =>
+  Object.values(groupBy(surveyResponses, sr => JSON.stringify(getDimensionCombo(sr, survey))));
+
+/**
+ * Not all changed survey responses need to trigger an outdated status update
+ * This method can be used to filter out changed records that are not worth
+ * considering, without having to look into the DB for more information
+ */
+const pushSurveyResponseChanges = changeDetails => {
+  const { type, new_record: newRecord, old_record: oldRecord } = changeDetails;
+
+  switch (type) {
+    case 'update': {
+      if (!oldRecord) {
+        // A new response has been created which may outdate an existing response
+        return [newRecord];
+      }
+
+      const records = [];
+      const fieldsOfInterest = ['data_time', 'end_time', 'survey_id', 'entity_id', 'id'];
+      if (!haveSameFields([newRecord, oldRecord], fieldsOfInterest)) {
+        // Updated record may have moved in a new dimension combo and thus may outdate
+        // an existing response
+        records.push(newRecord);
+        if (!oldRecord.outdated) {
+          // Updated record was the most recent response in its previous dimension combo,
+          // a new one may need to be selected
+          records.push(oldRecord);
+        }
+      } else if (isMarkedChange(changeDetails)) {
+        // Record was manually marked as changed, which may indicate that a client may want us to
+        // re-calculate its `outdated` status. In this case the old record is identical to the
+        // new, so no need to process it
+        records.push(newRecord);
+      }
+      return records;
     }
+    case 'delete':
+      // If the deleted record was the most recent response in its dimension combo,
+      // a new one must be selected
+      return oldRecord.outdated ? [] : [oldRecord];
+    default:
+      throw new Error(`${type} is not a supported change type`);
+  }
+};
+
+export class SurveyResponseOutdater extends ChangeHandler {
```

Comment:
The diff shown in this file is not very helpful, but most of the code here comes from https://github.com/beyondessential/tupaia/blob/3c47f0929be04e51ff2296403977d425f4b07a77/packages/database/src/changeHandlers/OutdatedStatusUpdater.js

Reply:
Thanks for pointing that out @kael89 - I diffed those locally and the changes are super minimal! Nice to collapse two files into one now the complexity lives in the parent class too :-)